plugins/Goals/DataTable/Filter/CalculateConversionPageRate.php

```diff
@@ -0,0 +1,113 @@
+{
+    /**
+     * Constructor.
+     *
+     * @param DataTable $table The table to eventually filter.
+     */
+    public function __construct($table)
+    {
+        parent::__construct($table);
+    }
+
+    /**
+     * @param DataTable $table
+     */
+    public function filter($table)
+    {
+
+        // Find all goal ids used in the table and store in an array
+        $goals = [];
+        foreach ($table->getRowsWithoutSummaryRow() as $row) {
+            if (isset($row[Metrics::INDEX_GOALS])) {
+                foreach ($row[Metrics::INDEX_GOALS] as $goalIdString => $metrics) {
+                    $goals[$goalIdString] = $goalIdString;
+                }
+            }
+        }
+
+        // Get the total top-level conversions for the goals in the table
+        $goalTotals = $this->getGoalTotalConversions($table, $goals);
+        if (count($goalTotals) === 0) {
+            return;
+        }
+
+        // Walk the rows and populate the nb_conversions_page_rate with nb_conversions_page_uniq / $goalTotals[goal id]
+        foreach ($table->getRowsWithoutSummaryRow() as &$row) {
+            if (isset($row[Metrics::INDEX_GOALS])) {
+                foreach ($row[Metrics::INDEX_GOALS] as $goalIdString => $metrics) {
+                    if (isset($row[Metrics::INDEX_GOALS][$goalIdString][Metrics::INDEX_GOAL_NB_CONVERSIONS_PAGE_UNIQ])) {
+
+                        $rate = Piwik::getQuotientSafe(
+                                $row[Metrics::INDEX_GOALS][$goalIdString][Metrics::INDEX_GOAL_NB_CONVERSIONS_PAGE_UNIQ],
+                                $goalTotals[$goalIdString],
+                                3
+                            );
+                        // Prevent page rates over 100% which can happen when there are subpages
+                        if ($rate > 1) {
+                            $rate = 1;
+                        }
+
+                        $row[Metrics::INDEX_GOALS][$goalIdString][Metrics::INDEX_GOAL_NB_CONVERSIONS_PAGE_RATE] = $rate;
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Get the conversions total for each goal in the top level datatable
+     *
+     * @param DataTable $table
+     * @param array $goalIds
+     * @return array
+     */
+    private function getGoalTotalConversions(DataTable $table, array $goalIds) : array
+    {
+
+        $goalTotals = [];
+
+        /** @var Site $site */
+        $site = $table->getMetadata('site');
+        if (empty($site)) {
+            return $goalTotals;
+        }
+        $idSite = $site->getId();
+
+        $period = $table->getMetadata('period');
+        $periodName = $period->getLabel();
+        $date = $period->getDateStart()->toString();
+
+        foreach ($goalIds as $idGoal => $g) {
+
+            $date = ($periodName === 'range' ? $date.','.$period->getDateEnd()->toString() : $date);
+            $archive = Archive::build($idSite, $periodName, $date);
```

Comment:
This one actually lags the segment as well. I'm wondering if it might be easier to move that to the Actions API. So we fetch the goal totals there and then pass the fetched totals to the filter call. That way it's a lot easier to access the required archive parameters and we don't need to rely on the datatable metadata.

Reply:
That makes sense, cleaner to have the goals totals injected into the filter than have the filter accessing archive data directly. I've committed an update, but I'm not sure that it will speed things up much.

Reply:
Actually I don't think this is going to work, `filterActionsDataTable` can accept a `DataTable\Map` as well as a `DataTable` so the goal retrieval will need to happen at the filter level since we may be dealing with multiple different datatables. I'll revert the change.